### ISSUE TITLE
wrong code at -O3 on x86_64-linux-gnu

### ISSUE THREAD

**Issue by zhendongsu** [llvm/llvm-project#49931]
|  |  |
| --- | --- |
| Bugzilla Link | [50587](https://llvm.org/bz50587) |
| Version | trunk |
| OS | All |

## Extended Description 
[611] % clangtk -v
clang version 13.0.0 (https://github.com/llvm/llvm-project.git 92ce29ee45b26513a9f4e42c6f287f43cb3de238)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/opfuzz/bin
Found candidate GCC installation: /usr/lib/gcc/i686-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
[612] % 
[612] % clangtk -O2 small.c; ./a.out
[613] % 
[613] % clangtk -O3 small.c
[614] % ./a.out
Aborted
[615] % 
[615] % cat small.c
```c
int a, b, c;
void f() {
  int d = 1, e = 1;
  while (1)
    for (; c < 1; c++) {
      if (!d && e)
        continue;
      if (a < 1) {
        if (e)
          b++;
        return;
      }
      d = 0;
      while (b)
        e++;
    }
}
int main() {
  f();
  if (b != 1)
    __builtin_abort ();
  return 0;
}
```

**Comment by fhahn** [comment]
Still reproduces on current `main`. Trying to bisect now.

**Comment by fhahn** [comment]
This is the same issue as #54023.

**Comment by fhahn** [comment]
Fixed by  4a699ae
